metrics_batcher.go

```diff
@@ -25,55 +25,97 @@ import (
 
 type metricsBatcher struct {
 	projectName string
-	allReqs     []*monitoringpb.CreateTimeSeriesRequest
 	allTss      []*monitoringpb.TimeSeries
 	allErrs     []error
-	// Counts all dropped TimeSeries by this exporter.
+
+	// Counts all dropped TimeSeries by this metricsBatcher.
 	droppedTimeSeries int
+
+	workers   []*worker
+	curWorker int
+	// respsChan is shared between metricsBatcher and worker goroutines.
+	respsChan chan *response
+
+	mc *monitoring.MetricClient
 }
 
-func newMetricsBatcher(projectID string) *metricsBatcher {
+func newMetricsBatcher(ctx context.Context, projectID string, numWorkers int, mc *monitoring.MetricClient) *metricsBatcher {
+	workers := make([]*worker, 0, numWorkers)
+	respsChan := make(chan *response, numWorkers)
+	for i := 0; i < numWorkers; i++ {
+		w := newWorker(ctx, mc, respsChan)
+		workers = append(workers, w)
+		go w.start()
+	}
 	return &metricsBatcher{
 		projectName:       fmt.Sprintf("projects/%s", projectID),
 		allTss:            make([]*monitoringpb.TimeSeries, 0, maxTimeSeriesPerUpload),
 		droppedTimeSeries: 0,
+		workers:           workers,
+		respsChan:         respsChan,
+		mc:                mc,
 	}
 }
 
-func (mb *metricsBatcher) recordDroppedTimeseries(numTimeSeries int, err error) {
+func (mb *metricsBatcher) recordDroppedTimeseries(numTimeSeries int, errs ...error) {
 	mb.droppedTimeSeries += numTimeSeries
-	mb.allErrs = append(mb.allErrs, err)
+	for _, err := range errs {
+		if err != nil {
+			mb.allErrs = append(mb.allErrs, err)
+		}
+	}
 }
 
 func (mb *metricsBatcher) addTimeSeries(ts *monitoringpb.TimeSeries) {
 	mb.allTss = append(mb.allTss, ts)
-	if len(mb.allTss) == maxTimeSeriesPerUpload {
-		mb.allReqs = append(mb.allReqs, &monitoringpb.CreateTimeSeriesRequest{
+	if len(mb.allTss) == maxTimeSeriesPerUpload && len(mb.workers) != 0 {
+		req := &monitoringpb.CreateTimeSeriesRequest{
 			Name:       mb.projectName,
 			TimeSeries: mb.allTss,
-		})
+		}
+		mb.workers[mb.curWorker].reqChan <- req
+		mb.nextWorker()
 		mb.allTss = make([]*monitoringpb.TimeSeries, 0, maxTimeSeriesPerUpload)
 	}
 }
 
-func (mb *metricsBatcher) export(ctx context.Context, mc *monitoring.MetricClient) {
-	// Last batch, if any.
-	if len(mb.allTss) > 0 {
-		mb.allReqs = append(mb.allReqs, &monitoringpb.CreateTimeSeriesRequest{
-			Name:       mb.projectName,
-			TimeSeries: mb.allTss,
-		})
+func (mb *metricsBatcher) nextWorker() {
+	mb.curWorker++
+	if mb.curWorker >= len(mb.workers) {
+		mb.curWorker = 0
 	}
+}
 
-	// Send create time series requests to Stackdriver.
-	for _, req := range mb.allReqs {
-		if err := createTimeSeries(ctx, mc, req); err != nil {
-			mb.recordDroppedTimeseries(len(req.TimeSeries), err)
+func (mb *metricsBatcher) close(ctx context.Context) error {
+	for _, w := range mb.workers {
+		w.stop()
+	}
+	for i := 0; i < len(mb.workers); i++ {
```

Comment:
here you need to poll num_requests_sent correct? I expect that one worker will add a response for every request pushed to the workers.